File: src/branchnexus/tmux/bootstrap.py
"""Tmux runtime bootstrap for selected WSL distribution."""

from __future__ import annotations

import logging as py_logging
import subprocess
from dataclasses import dataclass

from branchnexus.errors import BranchNexusError, ExitCode
from branchnexus.runtime.wsl_discovery import build_wsl_command

logger = py_logging.getLogger(__name__)


@dataclass
class BootstrapResult:
    tmux_available: bool
    install_attempted: bool = False


def _install_command_for_os_release(os_release: str) -> str:
    lowered = os_release.lower()
    if "debian" in lowered or "ubuntu" in lowered:
        return "sudo apt-get update && sudo apt-get install -y tmux"
    if "fedora" in lowered or "rhel" in lowered or "centos" in lowered:
        return "sudo dnf install -y tmux"
    if "arch" in lowered:
        return "sudo pacman -S --noconfirm tmux"
    if "opensuse" in lowered or "suse" in lowered:
        return "sudo zypper install -y tmux"
    raise BranchNexusError(
        "Unsupported distribution for automatic tmux install.",
        code=ExitCode.TMUX_ERROR,
        hint="Install tmux manually inside the selected distribution.",
    )


def _manual_install_guidance(os_release: str) -> str:
    try:
        cmd = _install_command_for_os_release(os_release)
        return f"Run this inside WSL: {cmd}"
    except BranchNexusError:
        return "Install tmux manually in the selected distribution and retry."


def ensure_tmux(
    distribution: str,
    *,
    auto_install: bool,
    runner: callable = subprocess.run,
) -> BootstrapResult:
    logger.debug("Checking tmux availability in distribution=%s", distribution)
    check = runner(
        build_wsl_command(distribution, ["command", "-v", "tmux"]),
        capture_output=True,
        text=True,
        check=False,
    )
    if check.returncode == 0:
        logger.debug("tmux is already installed in distribution=%s", distribution)
        return BootstrapResult(tmux_available=True, install_attempted=False)

    os_release_result = runner(
        build_wsl_command(distribution, ["cat", "/etc/os-release"]),
        capture_output=True,
        text=True,
        check=False,
    )
    os_release = os_release_result.stdout if os_release_result.returncode == 0 else ""
    logger.warning("tmux not found in distribution=%s auto_install=%s", distribution, auto_install)

    if not auto_install:
        logger.error("tmux missing and auto-install disabled in distribution=%s", distribution)
        raise BranchNexusError(
            "tmux is not installed in selected WSL distribution.",
            code=ExitCode.TMUX_ERROR,
            hint=_manual_install_guidance(os_release),
        )

    install_cmd = _install_command_for_os_release(os_release)
    logger.info("Attempting automatic tmux installation in distribution=%s", distribution)
    install_result = runner(
        build_wsl_command(distribution, ["bash", "-lc", install_cmd]),
        capture_output=True,
        text=True,
        check=False,
    )
    if install_result.returncode != 0:
        logger.error(
            "Automatic tmux installation failed in distribution=%s stderr=%s",
            distribution,
            install_result.stderr.strip(),
        )
        raise BranchNexusError(
            "Automatic tmux installation failed.",
            code=ExitCode.TMUX_ERROR,
            hint=_manual_install_guidance(os_release),
        )

    logger.info("Automatic tmux installation succeeded in distribution=%s", distribution)
    return BootstrapResult(tmux_available=True, install_attempted=True)

File: ts-src/tmux/bootstrap.ts
import { runCommand } from '../runtime/shell.js';
import { logger } from '../utils/logger.js';
import { BranchNexusError, ExitCode } from '../types/errors.js';
import { Platform, detectPlatform } from '../runtime/platform.js';

export interface BootstrapResult {
  tmuxAvailable: boolean;
  installAttempted: boolean;
}

const INSTALL_COMMANDS: Record<string, string> = {
  debian: 'sudo apt-get update && sudo apt-get install -y tmux',
  fedora: 'sudo dnf install -y tmux',
  arch: 'sudo pacman -S --noconfirm tmux',
  alpine: 'sudo apk add tmux',
  void: 'sudo xbps-install -Sy tmux',
  gentoo: 'sudo emerge app-misc/tmux',
  nixos: 'nix profile install nixpkgs#tmux',
  macos: 'brew install tmux',
};

function getInstallCommand(osRelease: string): string {
  const lowered = osRelease.toLowerCase();

  const debianIds = ['debian', 'ubuntu', 'pengwin', 'kali', 'mint', 'pop', 'elementary', 'zorin'];
  if (
    debianIds.some((name) => lowered.includes(name)) ||
    lowered.replace(/\s/g, '').includes('id_like=debian')
  ) {
    return INSTALL_COMMANDS['debian'];
  }

  const rhelIds = ['fedora', 'rhel', 'centos', 'rocky', 'almalinux', 'oracle', 'amazon'];
  if (rhelIds.some((name) => lowered.includes(name))) {
    return INSTALL_COMMANDS['fedora'];
  }

  const archIds = ['arch', 'manjaro', 'endeavouros', 'garuda'];
  if (
    archIds.some((name) => lowered.includes(name)) ||
    lowered.replace(/\s/g, '').includes('id_like=arch')
  ) {
    return INSTALL_COMMANDS['arch'];
  }

  if (lowered.includes('opensuse') || lowered.includes('suse')) {
    return 'sudo zypper install -y tmux';
  }

  if (lowered.includes('alpine')) {
    return INSTALL_COMMANDS['alpine'];
  }

  if (lowered.includes('void')) {
    return INSTALL_COMMANDS['void'];
  }

  if (lowered.includes('gentoo')) {
    return INSTALL_COMMANDS['gentoo'];
  }

  if (lowered.includes('nixos') || lowered.includes('nix')) {
    return INSTALL_COMMANDS['nixos'];
  }

  throw new BranchNexusError(
    'Unsupported distribution for automatic tmux install.',
    ExitCode.TMUX_ERROR,
    'Install tmux manually inside the selected distribution.'
  );
}

function getManualInstallGuidance(osRelease: string): string {
  try {
    const cmd = getInstallCommand(osRelease);
    return `Run this inside WSL: ${cmd}`;
  } catch {
    return 'Install tmux manually in the selected distribution and retry.';
  }
}

export async function ensureTmux(
  distribution: string,
  options?: { autoInstall?: boolean }
): Promise<BootstrapResult> {
  logger.debug(`Checking tmux availability in distribution=${distribution}`);

  const isWindows = detectPlatform() === Platform.WINDOWS;

  const checkCommand = isWindows
    ? ['wsl.exe', '-d', distribution, '--', 'command', '-v', 'tmux']
    : ['tmux'];

  try {
    const result = isWindows
      ? await runCommand(checkCommand)
      : await runCommand(['command', '-v', 'tmux']);

    if (result.exitCode === 0) {
      logger.debug(`tmux is already installed`);
      return { tmuxAvailable: true, installAttempted: false };
    }
  } catch {
    // Continue to install attempt
  }

  if (options?.autoInstall !== true) {
    throw new BranchNexusError(
      'tmux is not installed in selected WSL distribution.',
      ExitCode.TMUX_ERROR,
      'Install tmux manually or enable auto-install.'
    );
  }

  let osRelease = '';
  if (isWindows) {
    const osReleaseResult = await runCommand([
      'wsl.exe',
      '-d',
      distribution,
      '--',
      'cat',
      '/etc/os-release',
    ]);
    osRelease = osReleaseResult.stdout;
  } else if (detectPlatform() === Platform.MACOS) {
    osRelease = 'macos';
  } else {
    const fs = await import('node:fs/promises');
    try {
      osRelease = await fs.readFile('/etc/os-release', 'utf-8');
    } catch {
      osRelease = '';
    }
  }

  logger.warn(`tmux not found, attempting auto-install`);

  const installCmd = getInstallCommand(osRelease);

  // Try non-interactive install first
  const nonInteractiveCmd = installCmd.replace('sudo ', 'sudo -n ');
  logger.debug(`Trying non-interactive tmux install`);

  const nonInteractiveResult = isWindows
    ? await runCommand(['wsl.exe', '-d', distribution, '--', 'bash', '-lc', nonInteractiveCmd])
    : await runCommand(['bash', '-lc', nonInteractiveCmd]);

  if (nonInteractiveResult.exitCode === 0) {
    logger.info('Non-interactive tmux install succeeded');
    return { tmuxAvailable: true, installAttempted: true };
  }

  throw new BranchNexusError(
    'Interactive tmux installation required.',
    ExitCode.TMUX_ERROR,
    getManualInstallGuidance(osRelease)
  );
}
